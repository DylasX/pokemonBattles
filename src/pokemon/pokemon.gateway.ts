import {
  SubscribeMessage,
  MessageBody,
  ConnectedSocket,
  WebSocketGateway,
  OnGatewayInit,
  WebSocketServer,
  OnGatewayConnection,
  OnGatewayDisconnect,
} from '@nestjs/websockets';
import { Logger } from '@nestjs/common';
import { Socket, Server } from 'socket.io';
import { ConfigService } from '@nestjs/config';
import { PokemonService } from './pokemon.service';

@WebSocketGateway()
export class PokemonGateway
  implements OnGatewayInit, OnGatewayConnection, OnGatewayDisconnect
{
  @WebSocketServer() server: Server;

  private logger: Logger = new Logger('PokemonGateway');

  constructor(
    private configService: ConfigService,
    private pokemonService: PokemonService,
  ) {}

  @SubscribeMessage('joinRoom')
  async handleRoomJoin(
    @ConnectedSocket()
    client: Socket & Partial<Record<any, any>>,
    @MessageBody() data: any,
  ) {
    //just to access sids
    const prviousData: any = this.server.of('/').in(data.room);
    const clientsAdapter = Object.keys(
      Object.fromEntries(prviousData.adapter.sids),
    );

    if (data.room && data.pokemon && clientsAdapter.length < 2) {
      client.join(data.room);
      client.room = data.room;
      try {
        //Get general data
        const res = await this.pokemonService.getPokemon(
          data.pokemon?.id.toString(),
        );
        res.image = `${this.configService.get('REPO_IMAGES')}normal/${
          res.name
        }.gif`;
        client.pokemon = res;

        //Get data abilities
        if (data.pokemon.selectedMoves) {
          client.pokemon.moves = await Promise.all(
            this.pokemonService.generateMoves(data.pokemon.selectedMoves),
          );
        } else {
          client.pokemon.moves = await Promise.all(
            this.pokemonService.generateMoves(
              client.pokemon.moves
                .sort(() => Math.random() - Math.random())
                .slice(0, 4)
                .map((e) => e.move.name),
            ),
          );
        }

        client.broadcast.to(data.room).emit('newOpponent', {
          room: client.room,
          user: { pokemon: client.pokemon, userId: client.id },
        });

        return client.emit('joinedRoom', {
          room: client.room,
          user: { pokemon: client.pokemon, userId: client.id },
        });
      } catch (error) {
        console.error(error);
        this.logger.error(`Invalid pokemon with Id: ${data.pokemon}`);
      }
    } else {
      this.logger.error('Full room or body no supplied');
    }
  }

  @SubscribeMessage('previousPokemon')
  handlePreviousPokemon(client: Socket & Partial<Record<string, string>>) {
    const payload = {
      room: client.room,
      user: { pokemon: client.pokemon, userId: client.id },
    };
    client.broadcast.to(client.room).emit('previousOpponent', payload);
  }

  sleep(milliseconds) {
    return new Promise((resolve) => setTimeout(resolve, milliseconds));
  }

  @SubscribeMessage('leaveRoom')
  handleRoomLeave(client: Socket & Partial<Record<string, string>>) {
    this.server.to(client.room).emit('leftRoom', {
      room: client.room,
      pokemon: client.pokemon,
      userId: client.id,
    });
    client.disconnect(true);
  }

  disconnectAllSockets() {
    this.server.disconnectSockets();
  }

  afterInit(server: Server) {
    this.logger.log('Init');
  }

  handleDisconnect(client: Socket & Partial<Record<string, string>>) {
    this.logger.log(`Client disconnected: ${client.id}`);
    this.server.to(client.room).emit('leftRoom', {
      room: client.room,
      pokemon: client.pokemon,
      userId: client.id,
    });
    client.leave(client.room);
  }

  handleConnection(client: Socket, ...args: any[]) {
    this.logger.log(`Client connected: ${client.id}`);
  }
}
