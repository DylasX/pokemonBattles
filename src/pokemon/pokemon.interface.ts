export interface Pokemon {
  abilities: [];
  base_experience: number;
  forms: [];
  game_indices: [];
  height: number;
  held_items: [];
  id: number;
  image: string;
  is_default: boolean;
  location_area_encounters: string;
  moves: any[];
  name: string;
  order: number;
  past_types: [];
  species: Record<string, unknown>;
  sprites: Record<string, unknown>;
  stats: [];
  types: [];
  weight: number;
}

export interface PokemonMove {
  accuracy: number;
  id: number;
  pp: number;
  power: number;
  priority: number;
  target: Record<string, unknown>;
  type: Record<string, unknown>;
  stat_changes: any[];
  name: string;
}
