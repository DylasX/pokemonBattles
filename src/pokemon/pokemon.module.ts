import { Module, HttpModule } from '@nestjs/common';
import { PokemonGateway } from './pokemon.gateway';
import { PokemonService } from './pokemon.service';

@Module({
  imports: [HttpModule],
  controllers: [],
  providers: [PokemonGateway, PokemonService],
})
export class PokemonModule {}
