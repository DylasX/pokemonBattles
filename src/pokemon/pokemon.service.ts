import { Injectable, HttpService } from '@nestjs/common';
import { Pokemon } from './pokemon.interface';
import { AxiosResponse } from 'axios';
import { ConfigService } from '@nestjs/config';
import { map } from 'rxjs/operators';

@Injectable()
export class PokemonService {
  constructor(
    private http: HttpService,
    private configService: ConfigService,
  ) {}
  getPokemon(id: string): Promise<Pokemon> {
    return this.http
      .get<{ [data: string]: Pokemon }>(
        `${this.configService.get(`PUBLIC_API_URL`)}pokemon/${id}`,
      )
      .pipe(map((res: AxiosResponse) => res.data))
      .toPromise();
  }
}
