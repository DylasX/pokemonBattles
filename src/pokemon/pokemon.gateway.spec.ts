import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { AppModule } from '../app.module';
import { RedisIoAdapter } from '../adapters/redis.adapter';
import * as io from 'socket.io-client';
import { PokemonService } from '../pokemon/pokemon.service';
import { HttpModule } from '@nestjs/axios';

describe('AppController (e2e)', () => {
  let app: INestApplication;
  let connectToSocketIO: () => any;
  let pokemonService;
  const mockJoinRoom = {
    room: 'general',
    pokemon: {
      id: 1,
      selectedMoves: [1, 2, 3, 4],
    },
  };
  const mockPokemon = {
    abilities: [],
    base_experience: 1,
    forms: [],
    game_indices: [],
    height: 1,
    held_items: [],
    id: 1,
    image: 'x',
    is_default: true,
    location_area_encounters: 'x',
    moves: [],
    name: 'x',
    order: 1,
    past_types: [],
    species: {},
    sprites: {},
    stats: [],
    types: [],
    weight: 1,
  };

  const mockMoves = [
    {
      accuracy: 1,
      id: 1,
      pp: 1,
      power: 1,
      priority: 1,
      target: {},
      type: {},
      stat_changes: [],
      name: 'testMove',
    },
  ];

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule, HttpModule],
      providers: [PokemonService],
    }).compile();

    pokemonService = moduleFixture.get<PokemonService>(PokemonService);

    app = moduleFixture.createNestApplication();
    app.useWebSocketAdapter(new RedisIoAdapter(app));
    await app.init();
    await app.listen(4000);
    const httpServer = app.getHttpServer();
    connectToSocketIO = () =>
      io.connect(`http://127.0.0.1:${httpServer.address().port}`, {
        transports: ['websocket'],
      });
  });

  afterEach(async () => {
    await app.close();
  });

  it('should connect and disconnect', (done) => {
    const socket = connectToSocketIO();

    socket.on('connect', () => {
      socket.disconnect();
    });

    socket.on('disconnect', (reason) => {
      expect(reason).toBe('io client disconnect');
      done();
    });
    socket.on('error', done);
  });

  it('should join room', async (done) => {
    const socket = connectToSocketIO();

    //mock
    jest.spyOn(pokemonService, 'getPokemon').mockResolvedValue(mockPokemon);

    jest
      .spyOn(pokemonService, 'generateMoves')
      .mockReturnValue([Promise.resolve(mockMoves)]);

    socket.on('connect', () => {
      socket.emit('joinRoom', mockJoinRoom);
      socket.on('joinedRoom', (data) => {
        delete data.user.userId;
        expect(data).toMatchSnapshot();
        socket.disconnect();
      });
    });

    socket.on('disconnect', (reason) => {
      expect(reason).toBe('io client disconnect');
      done();
    });
    socket.on('error', done);
  });

  // it('should send previous pokemon', async (done) => {
  //   const promisifySocketHandler = (socket, event) =>
  //     new Promise((resolve) => socket.on(event, resolve));

  //   const allSocketsHandle = (sockets, event) =>
  //     sockets.map((socket) => promisifySocketHandler(socket, event));

  //   const sockets = [connectToSocketIO(), connectToSocketIO()];
  //   sockets.forEach((socket) => socket.on('error', done));

  //   await allSocketsHandle(sockets, 'connect');
  //   await sockets[0].emit('joinRoom');

  //   sockets[0].on('joinedRoom', async (data) => {
  //     await sockets[0].emit('previousPokemon');
  //   });
  // });
});
