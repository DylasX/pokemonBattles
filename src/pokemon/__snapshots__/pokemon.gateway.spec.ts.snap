// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) should join room 1`] = `
Object {
  "room": "general",
  "user": Object {
    "pokemon": Object {
      "abilities": Array [],
      "base_experience": 1,
      "forms": Array [],
      "game_indices": Array [],
      "height": 1,
      "held_items": Array [],
      "id": 1,
      "image": "https://pokeres.bastionbot.org/images/pokemon/1.png",
      "is_default": true,
      "location_area_encounters": "x",
      "moves": Array [
        Array [
          Object {
            "accuracy": 1,
            "id": 1,
            "name": "testMove",
            "power": 1,
            "pp": 1,
            "priority": 1,
            "stat_changes": Array [],
            "target": Object {},
            "type": Object {},
          },
        ],
      ],
      "name": "x",
      "order": 1,
      "past_types": Array [],
      "species": Object {},
      "sprites": Object {},
      "stats": Array [],
      "types": Array [],
      "weight": 1,
    },
  },
}
`;
